### ISSUE TITLE
Create a time histogram widget

### ISSUE THREAD

**Issue by Orbiter** [loklak/loklak_webclient#10]
The context of this feature is described in https://groups.google.com/d/msg/loklak/7azwyuslmTU/JXh0vrgXhSAJ

The framework page for this feature is created with https://github.com/loklak/loklak_webclient/issues/9
There should be a widget which shows the date histogram that is delivered by the loklak_server backend using calls like http://loklak.org/api/search.json?q=spacex%20since:2015-04-01_17:10%20until:2015-04-23_18:20&source=cache&count=0&fields=created_at,screen_name,mentions,hashtags&limit=6&timezoneOffset=-120

The histogramm data in the json of the call to the api can have steps: by-minute, by-hour or by-day. The histogram graph dimension names should reflect that. An easy approach to realize this can be morris.js to render the histogram. D3.js would be more flexible but also more difficult to use.

![message_evaluation_sketch](https://cloud.githubusercontent.com/assets/238730/7424468/b3fea152-efa1-11e4-8c6d-8bc85c093178.png)

A very nice feature would be a date slider at the border of the histogram to move the date zones, i.e. shrink them. This would make it possible to omit the time input fields for the search page.

**Comment by Orbiter** [comment]
Thats the code I used for the time histogram in YaCy, maybe this helps. Feel free to copy from it:

<link rel="stylesheet" href="/env/morris.css">
<script src="/js/raphael-min.js"></script>
<script src="/js/morris.js"></script>
<script>
	document.getElementById("datehistogram").style = "height:200px";
	dates_in_content_dts = [#{element}#"#[name]#","#[count]#"#(nl)#::,#(/nl)##{/element}#];
	var parsed = [];
    for (var i = 0; i < dates_in_content_dts.length; i = i + 2) {
        var date = dates_in_content_dts[i];
        var count = dates_in_content_dts[i + 1];
        if (date && count) {parsed[parsed.length] = {x: date,y: count};};
    };
    if (parsed.length > 0) Morris.Bar({
      element: 'datehistogram',
      data: parsed,
      xkey: 'x',
      ykeys: ['y'],
      labels: ['number of documents about this date'],
      yLabelFormat: function (y) { return y.toString() + ' docs'; },
      barColors: function (row, series, type) {
    	  var d = new Date(row.label);
          if (d.getDay() === 6) return '#4aaf46'; //saturday
          if (d.getDay() === 0) return '#4aaf46'; //sunday
    	  return '#3574c0';
      },
      hideHover: 'false'
    }).on('click', function(i, row) {
    	var query = document.getElementsByClassName('searchinput')[0].getAttribute("value");
    	var onp = -1, fromp = -1, top = -1;
    	if ((onp = query.indexOf("on:")) >= 0) {
    		query = query.substring(0, onp - 1);
    	}
    	if ((fromp = query.indexOf("from:")) < 0) {
    		query = query + " from:" + row.x;
    		document.getElementsByClassName('searchinput')[0].value = query;
    		document.getElementById('Enter').click();
    	} else if ((top = query.indexOf("to:")) < 0) {
    		query = query + " to:" + row.x;
    		document.getElementsByClassName('searchinput')[0].value = query;
    		document.getElementById('Enter').click();
    	} else {
    		query = query.substring(0, fromp) + " on:" + row.x;
    		document.getElementsByClassName('searchinput')[0].value = query;
    		document.getElementById('Enter').click();
    	}
    	var date = row.x;
    	console.log(i, row, query);
    });
</script>

**Comment by Orbiter** [comment]
who is working on this?

**Comment by aneeshd16** [comment]
@Orbiter I am working on this.

**Comment by Orbiter** [comment]
Thats already great, I just tweetet a screenshot: https://twitter.com/0rb1t3r/status/594946162530111488

- what is not needed: time selection on 5-minute basis. That was a misunderstanding. Selection on hourly basis is sufficient (and good!)
- what could be good: subtitels under the histogram for more times so it's better to see in screenshots that the horizontal scale is time

**Comment by Orbiter** [comment]
I posted an update on twitter: https://twitter.com/0rb1t3r/status/595293466957094912
What is not yet working: if you use the shared link, the Date Fields say "Invalid Date".
- Can you move the date from the url into the select field?
- can you enlarge the date field with a bit so the date is better visible? Don't make them too wide so the input line aligns with the histogram

**Comment by aneeshd16** [comment]
@Orbiter its working fine on my end. For example, this share link: http://localhost:3001/statistics?q=youtube&since=2015-04-01_01:00&until=2015-05-01_06:00 works, the correct dates automatically come in the date boxes.
Will enlarge the date fields so that the date is visible.

**Comment by Orbiter** [comment]
Ah, looks like a browser dependency! I am using firefox - does not work. Tested this with chrome: works!
Safari: does NOT work.

Please test again with firefox (and maybe safari, if you have a mac)

**Comment by aneeshd16** [comment]
That was unexpected. Thanks! Will test.

**Comment by Orbiter** [comment]
here is another problem: I was able to run the webclient on loklak.net:

http://loklak.net:3001/statistics?q=rp15&since=2015-05-01_00:00&until=2015-05-07_23:00

However, if you call the url, the recommended community-url has still the localhost path in it:
http://localhost:3001/statistics?q=rp15&since=2015-05-01_00:00&until=2015-05-07_23:00

**Comment by aneeshd16** [comment]
You need the change the `domain` environment variable in `custom_configFile.json` to loklak.net.
I just committed some changes in b1865d936ba566eec58f7aa0dd0c5a6cf47b4c7b, fixed the "Invalid Date" issue and stretched the input boxes to show the full date.
Tested with Chrome, Firefox and even IE. Should work with Safari as the part that resolves the dates is now not dependent on the browser, but on moment.js

**Comment by Orbiter** [comment]
ah yes, my fault. fixed it. your fix for the date also worked! Thank you!

What I still wish, is the url in the url line. I know that this somehow the angular.js/node.js style of doing things. If you make the 'Search' button into a regular http get, that should already work through re-loading. Bad idea?

**Comment by aneeshd16** [comment]
I wish for that too! Spent a lot of time on getting the url to change without reloading, before finally committing whatever changes i had and promising myself to look for a solution. A solution definitely exists, we are just yet to find it. It could turn out to be very simple.

I thought about making the 'Search' button into a regular http get, but it kinda defeats the purpose of having a solid AJAX setup with angular, and plus slows the user experience.

Hoping to find the solution soon!

**Comment by Orbiter** [comment]
args! I deployed the latest change on loklak.net and the histogram is missing!

```
Leerer String an getElementById() übergeben. jquery.min.js:2:0
Array [ Array[3], Array[3], Array[3], Array[3], Array[3], Array[3], Array[3], Array[3], Array[3], Array[3], 90 weitere… ] main.js:522:4
"setting chart values" main.js:643:20
"Error: n is undefined
t.Grid</r.prototype.setData/this.data</a.y<@http://cdn.oesmith.co.uk/morris-0.4.3.min.js:1:3586
t.Grid</r.prototype.setData/this.data<@http://cdn.oesmith.co.uk/morris-0.4.3.min.js:1:3527
t.Grid</r.prototype.setData@http://cdn.oesmith.co.uk/morris-0.4.3.min.js:1:3169
link/<@http://loklak.net:3001/js/main.js:644:21
$CompileProvider/this.$get</Attributes.prototype.$set/<@http://loklak.net:3001/js/main.js:12903:13
forEach@http://loklak.net:3001/js/main.js:6389:11
$CompileProvider/this.$get</Attributes.prototype.$set@http://loklak.net:3001/js/main.js:12901:1
interpolateFnWatchAction@http://loklak.net:3001/js/main.js:14223:23
interpolateFnWatcher@http://loklak.net:3001/js/main.js:16282:17
watchGroupAction@http://loklak.net:3001/js/main.js:20046:13
$RootScopeProvider/this.$get</Scope.prototype.$digest@http://loklak.net:3001/js/main.js:20374:23
$RootScopeProvider/this.$get</Scope.prototype.$apply@http://loklak.net:3001/js/main.js:20637:13
done@http://loklak.net:3001/js/main.js:15764:36
completeRequest@http://loklak.net:3001/js/main.js:15954:7
createHttpBackend/</jsonpDone<@http://loklak.net:3001/js/main.js:15864:9
jsonpReq/callback@http://loklak.net:3001/js/main.js:15985:9
" main.js:17721:17

"morris resized" main.js:625:20
TypeError: r is undefined morris-0.4.3.min.js:1:6267
TypeError: r is undefined morris-0.4.3.min.js:1:26589

```
tried with http://loklak.net:3001/statistics?q=rp15&since=2015-05-01_00:00&until=2015-05-07_23:00

**Comment by aneeshd16** [comment]
@Orbiter I have tried to fix in the latest commit, but haven't been able to test it.
Also, do not deploy all files to the server. We are not using node yet, and our final output is just a few static HTML/CSS/JS files. Run `gulp prod`, and just deploy the contents of the `build` directory onto the server.

**Comment by Orbiter** [comment]
that did not work. If I open index.html then nothing happens, not even an error in firefox debugger. Is it necessary to host the  build directory or can I open it as file?

**Comment by Orbiter** [comment]
One more point: can you print out the total number of all messages (= sum it up) of the histogram? maybe somewhere below, but not in an extra line to save space. Like
2015-05-05 14:59  Tweet Count: 42
Total Tweet Count 2015-05-05 14:59 - 2015-05-05 15:00: 734

**Comment by sudheesh001** [comment]
@Orbiter The build/* directory needs to be hosted using some kind of an http server, like 
`python -m SimpleHTTPServer` or `http-server -o` or Apache :+1:

**Comment by Orbiter** [comment]
I tried that but it did not work! loklak.net is now alive with the gulp dev environment.

**Comment by Orbiter** [comment]
I believe we can consider this feature as implemented and you all did a good job working together as a team. This is fun! Thank you!

**Comment by aneeshd16** [comment]
Michael, using the `gulp dev` environment won't work because of browser sync. It wont be able to handle multiple clients. You can try opening it in two different tabs and you'll see that any thing you type will be synced across to the other tab.

We need a web server for this. I'll search for a concrete solution for this.

**Comment by aneeshd16** [comment]
@Orbiter I have set it up temporarily here: http://loklak.parseapp.com It can also be redirected to loklak.net. I've used a node environment for this which rewrites all incoming urls to the index.html file.

Setting up an angular application requires a web server that does URL rewriting manually OR some sort of automatic URL rewriting configuration on the server. This should help: http://stackoverflow.com/a/17665635/2463357
Hope it works!
